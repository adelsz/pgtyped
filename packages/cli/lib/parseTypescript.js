import { parseTSQuery } from '@pgtyped/parser';
import ts from 'typescript';
export function parseFile(sourceFile, transformConfig) {
    const foundNodes = [];
    parseNode(sourceFile);
    function parseNode(node) {
        if ((transformConfig === null || transformConfig === void 0 ? void 0 : transformConfig.mode) === 'ts-implicit' &&
            node.kind === ts.SyntaxKind.CallExpression) {
            const callNode = node;
            const functionName = callNode.expression.getText();
            if (functionName === transformConfig.functionName) {
                const queryName = callNode.parent.getChildren()[0].getText();
                const queryText = callNode.arguments[0].getText().slice(1, -1).trim();
                foundNodes.push({
                    queryName,
                    queryText,
                });
            }
        }
        if (node.kind === ts.SyntaxKind.TaggedTemplateExpression) {
            const queryName = node.parent.getChildren()[0].getText();
            const taggedTemplateNode = node;
            const tagName = taggedTemplateNode.tag.getText();
            const queryText = taggedTemplateNode.template
                .getText()
                .replace('\n', '')
                .slice(1, -1)
                .trim();
            if (tagName === 'sql') {
                foundNodes.push({
                    queryName,
                    queryText,
                });
            }
        }
        ts.forEachChild(node, parseNode);
    }
    const queries = [];
    const events = [];
    for (const node of foundNodes) {
        const { query, events: qEvents } = parseTSQuery(node.queryText, node.queryName);
        queries.push(query);
        events.push(...qEvents);
    }
    return { queries, events };
}
export const parseCode = (fileContent, fileName = 'unnamed.ts', transformConfig) => {
    const sourceFile = ts.createSourceFile(fileName, fileContent, ts.ScriptTarget.ES2015, true);
    return parseFile(sourceFile, transformConfig);
};
//# sourceMappingURL=parseTypescript.js.map