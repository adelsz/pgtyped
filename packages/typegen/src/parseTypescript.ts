import { ParseEvent, parseTSQuery, TSQueryAST } from '@pgtyped/parser';
import ts from 'typescript';
import { TransformMode } from './config.js';

interface INode {
  queryName: string;
  queryText: string;
}

export type TSParseResult = { queries: TSQueryAST[]; events: ParseEvent[] };

export function parseFile(
  sourceFile: ts.SourceFile,
  mode: TransformMode | undefined,
): TSParseResult {
  const foundNodes: INode[] = [];
  parseNode(sourceFile);

  function parseNode(node: ts.Node) {
    if (
      mode?.mode === 'ts-implicit' &&
      node.kind === ts.SyntaxKind.CallExpression
    ) {
      const callNode = node as ts.CallExpression;
      const functionName = callNode.expression.getText();
      if (functionName === mode.functionName) {
        const queryName = callNode.parent.getChildren()[0].getText();
        const queryText = callNode.arguments[0].getText().slice(1, -1).trim();
        foundNodes.push({
          queryName,
          queryText,
        });
      }
    }

    if (node.kind === ts.SyntaxKind.TaggedTemplateExpression) {
      const queryName = node.parent.getChildren()[0].getText();
      const taggedTemplateNode = node as ts.TaggedTemplateExpression;
      const tagName = taggedTemplateNode.tag.getText();
      const queryText = taggedTemplateNode.template
        .getText()
        .replace('\n', '')
        .slice(1, -1)
        .trim();
      if (tagName === 'sql') {
        foundNodes.push({
          queryName,
          queryText,
        });
      }
    }

    ts.forEachChild(node, parseNode);
  }

  const queries: TSQueryAST[] = [];
  const events: ParseEvent[] = [];
  for (const node of foundNodes) {
    const { query, events: qEvents } = parseTSQuery(
      node.queryText,
      node.queryName,
    );
    queries.push(query);
    events.push(...qEvents);
  }

  return { queries, events };
}

export const parseCode = (
  fileContent: string,
  fileName = 'unnamed.ts',
  mode?: TransformMode,
) => {
  const sourceFile = ts.createSourceFile(
    fileName,
    fileContent,
    ts.ScriptTarget.ES2015,
    true,
  );
  return parseFile(sourceFile, mode);
};
