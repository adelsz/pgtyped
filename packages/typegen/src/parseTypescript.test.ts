import { parseCode } from './parseTypescript.js';

test('parser finds string template in correct file', () => {
  const fileContent = `
    const sql : any = null;

    const query = sql\`
      select id, name, age from users;
    \`;
  `;

  const result = parseCode(fileContent);
  expect(result).toMatchSnapshot();
});

test('parser finds string template in incorrect file', () => {
  const fileContent = `
    const sql  ny =/ null;

    const query = sql\`
      select id, name, age from users;
    \`;
  `;

  const result = parseCode(fileContent);
  expect(result).toMatchSnapshot();
});
