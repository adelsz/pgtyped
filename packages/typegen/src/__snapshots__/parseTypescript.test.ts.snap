// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser finds string template in correct file 1`] = `
Object {
  "events": Array [],
  "queries": Array [
    Object {
      "name": "query",
      "params": Array [],
      "text": "select id, name, age from users",
    },
  ],
}
`;

exports[`parser finds string template in incorrect file 1`] = `
Object {
  "events": Array [],
  "queries": Array [
    Object {
      "name": "query",
      "params": Array [],
      "text": "select id, name, age from users",
    },
  ],
}
`;
